cmake: increase minimal required version to 3.20.0

tejlmand opened the pull request:
Move to CMake 3.20.0.

At the Toolchain WG it was decided to move to CMake 3.20.0.

The main reason for increasing CMake version is better toolchain
support.
    
Better toolchain support is added in the following CMake versions:
- armclang, CMake 3.15
- Intel oneAPI, CMake 3.20
- IAR, CMake 3.15 and 3.20
    

See MoM: https://docs.google.com/document/d/1IQKBK-GcJNZG0O9QArqYfvb6Huk5xHscN-XIGEZr-z8/edit#heading=h.gc6zh884razh


Note: ~DNM because the PR contains an unmerged manifest update.~
See: https://github.com/zephyrproject-rtos/lvgl/pull/13 <-- is merged and manifest is updated to SHA revision.

tejlmand committed:
cmake: increase minimal required version to 3.20.0

Move to CMake 3.20.0.

At the Toolchain WG it was decided to move to CMake 3.20.0.

The main reason for increasing CMake version is better toolchain
support.

Better toolchain support is added in the following CMake versions:
- armclang, CMake 3.15
- Intel oneAPI, CMake 3.20
- IAR, CMake 3.15 and 3.20

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
Revert "cmake: manual handling of CMakeCheckCompilerFlag.cmake inclusion"

With the change to cmake_minimum_required(VERSION 3.20.0) then this
commit is no longer needed.

This reverts commit 6ca2bf25cb1da64edef1fe7592ab6fccdd4a16f7.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
PR to update the CMake version in docker is found here: https://github.com/zephyrproject-rtos/docker-image/pull/71.

That should fix the CI issue regarding old CMake version in use.

gmarull reviewed (changes requested):
nit: https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/getting_started/index.rst needs to be updated with new minimum version

mbolivar-nordic reviewed (changes requested):
:disappointed: 

I agree with @gmarull, but this is not just a nit.

The latest Ubuntu LTS is shipping an incompatible version:

```
$ apt-cache show cmake | grep Version
Version: 3.16.3-1ubuntu1
```

Zephyr v1.14 LTS documentation has a detailed HOWTO on how to get cmake from various different sources to get something more recent than the distribution version:

https://docs.zephyrproject.org/1.14.1/getting_started/installation_linux.html#install-requirements-and-dependencies

I would expect something similar to be resurrected here.

galak commented:
@tejlmand you can use docker image v0.18.2 now.

tejlmand commented:
> Zephyr v1.14 LTS documentation has a detailed HOWTO on how to get cmake from various different sources to get something more recent than the distribution version:
> 
> https://docs.zephyrproject.org/1.14.1/getting_started/installation_linux.html#install-requirements-and-dependencies
> 
> I would expect something similar to be resurrected here.

It already exists:
https://docs.zephyrproject.org/latest/getting_started/installation_linux.html#installation-linux

> A recent CMake version is required. Check what version you have by using cmake --version. If you have an older version, there are several ways of obtaining a more recent one:
>
> -  On Ubuntu, you can follow the instructions for adding the kitware third-party apt repository to get an updated version of cmake using apt.

But I will add more description on the getting started page.

tejlmand committed:
docs: Update docs for CMake 3.20.0 minimal version

Update docs to reference CMake 3.20.0 as minimum required version.

Added Kitware APT repository to the getting started page.

Updated installation Linux download descriptions from CMake 3.15.3 to
CMake 3.21.1 as that is the version currently being installed when
using the Kitware APT repository.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
> nit: https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/getting_started/index.rst needs to be updated with new minimum version

Done.

> But I will add more description on the getting started page.

and done.

evgeniy-paltsev commented:
Just FYI: we use 3.20.2 cmake in our internal verification - so this change is OK for ARC.

tejlmand commented on doc/getting_started/index.rst:
```diff
@@ -76,6 +76,15 @@ The current minimum required version for the main dependencies are:
+
+         .. code-block:: bash
+
+            wget https://apt.kitware.com/kitware-archive.sh
+	    # Inspect that script is safe and execute
+	    sudo bash kitware-archive.sh
```
good point, no reason to mention it twice.

tejlmand replied:
done

tejlmand commented on doc/getting_started/index.rst:
```diff
@@ -95,6 +104,12 @@ The current minimum required version for the main dependencies are:
+	    It is also possible to execute each step in ``kitware-archive.sh``
+	    manually, see `kitware third-party apt repository <https://apt.kitware.com/>`_
+	    for detailed instructions.
+
```
we could shorten the note to something like:
> A detailed explanation of ``kitware-archive.sh`` can be found here `kitware third-party apt repository` <https://apt.kitware.com/>

tejlmand replied:
Or even move it up, and remove the note, like:

> Download, inspect and execute the Kitware archive script to add the Kitware APT repository to your sources list.
> A detailed explanation of kitware-archive.sh can be found here [kitware third-party apt repository](https://apt.kitware.com/)
> ```
> wget https://apt.kitware.com/kitware-archive.sh
> sudo bash kitware-archive.sh
> ```

gmarull replied:
Yep, sounds good:

> Or even move it up, and remove the note, like:
> 
> > Download, inspect and execute the Kitware archive script to add the Kitware APT repository to your sources list.
> > A detailed explanation of kitware-archive.sh can be found here [kitware third-party apt repository](https://apt.kitware.com/)
> > ```
> > wget https://apt.kitware.com/kitware-archive.sh
> > sudo bash kitware-archive.sh
> > ```

tejlmand replied:
done

tejlmand commented on doc/getting_started/index.rst:
```diff
@@ -76,6 +76,15 @@ The current minimum required version for the main dependencies are:
       .. _install_dependencies_ubuntu:
 
+      #. Download, inspect and execute the Kitware archive script to add the
+         Kitware APT repository to yoursources list::
```
done

tejlmand commented on doc/getting_started/index.rst:
```diff
@@ -76,6 +76,16 @@ The current minimum required version for the main dependencies are:
       .. _install_dependencies_ubuntu:
 
+      #. Download, inspect and execute the Kitware archive script to add the
+         Kitware APT repository to your sources list::
```
I was unsure :man_shrugging: so I just kept them.
Do you prefer them to be removed completely, or moved to the line below ?

gmarull replied:
`::` are used to introduce literals, see https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#literal-blocks

tejlmand replied:
I had the impression it was simply to get a single `:` at the end, but I see that is also the result.
> The handling of the :: marker is smart:
> - ...
> - If it is preceded by non-whitespace, the marker is replaced by a single colon.

Have moved in to line below.

tejlmand committed:
manifest: update lvgl for CMakeLists.txt fixes for CMake versions >=3.20

CMake >=3.20 requires file extensions explicitly added to source files.

See CMP0115:
> Starting in CMake 3.20, CMake prefers all source files to have their
> extensions explicitly listed:

This manifest update pulls in update for the lvgl module adding missing
extension.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
Buildkite failure is due to timeouts because of the amount of files changed --> CI runs for long time.

mbolivar-nordic commented:
@tejlmand consider sending email when this is merged so people know they need to update.
